File: android/src/main/java/com/dynatrace/android/agent/DynatraceRNBridge.java
package com.dynatrace.android.agent;

import static com.dynatrace.android.agent.DynatraceUtils.toHashMap;

import com.dynatrace.android.agent.conf.DynatraceConfigurationBuilder;
import com.dynatrace.android.agent.crash.PlatformType;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMapKeySetIterator;

import java.net.MalformedURLException;
import java.util.Hashtable;
import java.util.HashMap;
import java.util.Map;

import android.location.Location;

import org.json.JSONObject;

import com.dynatrace.android.agent.conf.DataCollectionLevel;
import com.dynatrace.android.agent.conf.UserPrivacyOptions;

public class DynatraceRNBridge extends ReactContextBaseJavaModule {
    
    private HashMap<String, WebRequestTiming> webTimings;
    private HashMap<String, DTXAction> actions;
    private static final String PLATFORM_ANDROID = "android";
    private static final String PLATFORM_IOS = "ios";
    private static final String DATA_COLLECTION_OFF = "OFF";
    private static final String DATA_COLLECTION_PERFORMANCE = "PERFORMANCE";
    private static final String DATA_COLLECTION_USERBEHAVIOR = "USER_BEHAVIOR";

    private static PrivateDTBridge _internal;

    public DynatraceRNBridge(ReactApplicationContext reactContext, PrivateDTBridge internal) {
        super(reactContext);
        webTimings = new HashMap<>();
        actions = new HashMap<>();
        _internal = internal;
    }

    @Override
    public String getName(){
        return "DynatraceBridge";
    }

    @Override
    public Map<String, Object> getConstants() {
        final Map<String, Object> constants = new HashMap<>();
        constants.put("PLATFORM_ANDROID", PLATFORM_ANDROID);
        constants.put("PLATFORM_IOS", PLATFORM_IOS);
        constants.put("DATA_COLLECTION_OFF", DATA_COLLECTION_OFF);
        constants.put("DATA_COLLECTION_PERFORMANCE", DATA_COLLECTION_PERFORMANCE);
        constants.put("DATA_COLLECTION_USERBEHAVIOR", DATA_COLLECTION_USERBEHAVIOR);
        return constants;
    }

    @ReactMethod
    public void start(ReadableMap configuration) throws Exception {
        if (configuration == null) {
            throw new Exception("Configuration is empty for manual startup! This is not possible");
        } else {
            if(configuration.getString("applicationId") == null || configuration.getString("beaconUrl") == null){
                throw new Exception("applicationId and beaconUrl can't be empty!");
            }

            DynatraceConfigurationBuilder builder = new DynatraceConfigurationBuilder(configuration.getString("applicationId"),
                configuration.getString("beaconUrl"));

            builder.withUserOptIn(configuration.getBoolean("userOptIn"));
            builder.withCrashReporting(configuration.getBoolean("reportCrash"));
            // 0 == Debug / 1 == Info
            builder.withDebugLogging(configuration.getInt("logLevel") == 0);

            Dynatrace.startup(getReactApplicationContext(), builder.buildConfiguration());
        } 
    }

    //
    // Expects a key which is generated in JS. This is to circumvent the async callback system.
    //
    @ReactMethod
    public void enterAction(String name, String key, String platform) {
        if (this.shouldWorkOnAndroid(platform)) {
            try {
                newAction(name, key);
            }
            catch (Exception e) {}
        }
    }

    @ReactMethod
    public void enterManualAction(String name, String key, String platform) {
        if (this.shouldWorkOnAndroid(platform)) {
			actions.put(key, Dynatrace.enterAction(name));
        }
    }

    @ReactMethod
    public void enterManualActionWithParent(String name, String key, String parentKey, String platform) {
		if (this.shouldWorkOnAndroid(platform)) {
			DTXAction parent = actions.get(parentKey);

			if (parent != null) {
				actions.put(key, Dynatrace.enterAction(name, parent));
			} else {
				enterManualAction(name, key, platform);
			}
		}
	}

    @ReactMethod
    public void leaveAction(String key, boolean leave, String platform) {
        if (this.shouldWorkOnAndroid(platform)) {
            DTXAction action = getAction(key);
            if (action == null) return;
            if (action instanceof DTXAutoAction){
                ((DTXAutoAction) action).startTimer();
            }else{
                action.leaveAction();
            }

            actions.remove(key);
        }
    }

    @ReactMethod
    public void cancelAction(String key, String platform) {
        if (this.shouldWorkOnAndroid(platform)) {
            DTXAction action = getAction(key);
            if (action == null) return;
            action.cancel();
            actions.remove(key);
        }
    }

    @ReactMethod
    public void endVisit(String platform) {
        if (this.shouldWorkOnAndroid(platform)) {
            Dynatrace.endVisit();
        }
    }

    @ReactMethod
    public void reportErrorWithoutStacktrace(String errorName, int errorCode, String platform) {
        if (this.shouldWorkOnAndroid(platform)) {
            Dynatrace.reportError(errorName, errorCode);
        }
    }

    /**
     * Reports a stacktrace
     * @param errorName Name of the Error - SyntaxError
     * @param errorValue Value of the Error
     * @param reason Reason for the Error
     * @param stacktrace Whole Stacktrace
     * @param platform Platform wise or both
     */
    @ReactMethod
    public void reportError(String errorName, String errorValue, String reason, String stacktrace, String platform){
        if (this.shouldWorkOnAndroid(platform)) {
            Dynatrace.reportError(PlatformType.CUSTOM, errorName, errorValue, reason, stacktrace);
        }
    }

    @ReactMethod
    public void reportCrash(String errorName, String reason, String stacktrace, boolean isRealError, boolean newSession, String platform){
        if (this.shouldWorkOnAndroid(platform)) {
            Dynatrace.reportCrash(isRealError ? PlatformType.JAVA_SCRIPT : PlatformType.CUSTOM, errorName, reason, stacktrace);
            if(newSession){
                Dynatrace.createNewSession();
            }     
        }
    }

    @ReactMethod
    public void reportErrorInAction(String key, String errorName, int errorCode, String platform) {
        if (this.shouldWorkOnAndroid(platform)) {
            DTXAction action = getAction(key);
            if (action == null) return;
            action.reportError(errorName, errorCode);
        }
    }

    @ReactMethod
    public void reportValue(String key, String valueName, String value, String platform) {
        if (this.shouldWorkOnAndroid(platform)) {
            DTXAction action = getAction(key);
            if (action == null) return;
            action.reportValue(valueName, value);
        }
    }

    @ReactMethod
    public void getRequestTag(String key, String url, Promise promise){
        DTXAction action = getAction(key);

        if (action == null){
            promise.resolve(Dynatrace.getRequestTag());
            return;
        };

        promise.resolve(action.getRequestTag());
    }

    @ReactMethod
    public void startWebRequestTiming(String requestTag, String url) {
        if (requestTag != null) {
            WebRequestTiming timing = Dynatrace.getWebRequestTiming(requestTag);
            if(timing != null){
                webTimings.put(requestTag, timing);
                timing.startWebRequestTiming();
            }
        }
    }
    
    @ReactMethod
    public void stopWebRequestTiming(String requestTag, String url, int responseCode, String responseMessage) {
        if (requestTag != null) {
            WebRequestTiming timing = webTimings.get(requestTag);
            if (timing != null) {
                try {
                    timing.stopWebRequestTiming(url, responseCode, responseMessage);
                    webTimings.remove(requestTag);
                } catch (MalformedURLException ex) {
                    // do nothing
                }
            }
        }
    }

    @ReactMethod
    public void identifyUser(String user, String plaform) {
        if (this.shouldWorkOnAndroid(plaform)) {
            Dynatrace.identifyUser(user);
        }
    }

    @ReactMethod
    public void reportEventInAction(String actionKey, String name, String platform) {
        if (this.shouldWorkOnAndroid(platform)) {
            DTXAction action = getAction(actionKey);
            if (action == null) return;
            action.reportEvent(name);
        }
    }

    @ReactMethod
    public void reportStringValueInAction(String actionKey, String name, String value, String platform) {
        if (this.shouldWorkOnAndroid(platform)) {
            DTXAction action = getAction(actionKey);
            if (action == null) return;
            action.reportValue(name, value);
        }
    }

    @ReactMethod
    public void reportIntValueInAction(String actionKey, String name, int value, String platform) {
        if (this.shouldWorkOnAndroid(platform)) {
            DTXAction action = getAction(actionKey);
            if (action == null) return;
            action.reportValue(name, value);
        }
    }

    @ReactMethod
    public void reportDoubleValueInAction(String actionKey, String name, double value, String platform) {
        if (this.shouldWorkOnAndroid(platform)) {
            DTXAction action = getAction(actionKey);
            if (action == null) return;
            action.reportValue(name, value);
        }
    }

    @ReactMethod
    public void sendEvent(String name, ReadableMap attributes, String platform) {
        if(this.shouldWorkOnAndroid(platform)) {
            Dynatrace.sendEvent(name, new JSONObject(toHashMap(attributes)));
        }
    }

    @ReactMethod
    public void sendBizEvent(String type, ReadableMap attributes, String platform) {
        if(this.shouldWorkOnAndroid(platform)) {
            Dynatrace.sendBizEvent(type, new JSONObject(toHashMap(attributes)));
        }
    }

    @ReactMethod
    public void setGPSLocation(double latitude, double longitude, String platform){
        if(this.shouldWorkOnAndroid(platform)){
            Location location = new Location("");
            location.setLatitude(latitude);
            location.setLongitude(longitude);
            Dynatrace.setGpsLocation(location);
        }
    }

    @ReactMethod
    public void flushEvents(String platform){
        if(this.shouldWorkOnAndroid(platform)){
            Dynatrace.flushEvents();
        }
    }

    @ReactMethod
    public void isCrashReportingOptedIn(String platform, Promise promise){
        if(this.shouldWorkOnAndroid(platform)){
            promise.resolve(Dynatrace.isCrashReportingOptedIn());
        }
    }

    @ReactMethod
    public void setCrashReportingOptedIn(boolean crashReporting, String platform){
        if(this.shouldWorkOnAndroid(platform)){
            Dynatrace.setCrashReportingOptedIn(crashReporting);
        }
    }

    @ReactMethod
    public void setDataCollectionLevel(String collectionLevel, String platform){
        if(this.shouldWorkOnAndroid(platform)){
            Dynatrace.setDataCollectionLevel(DataCollectionLevel.valueOf(collectionLevel));
        }
    }

    @ReactMethod
    public void getDataCollectionLevel(String platform, Promise promise){
        if(this.shouldWorkOnAndroid(platform)){
            DataCollectionLevel level = Dynatrace.getDataCollectionLevel();
            promise.resolve(level.name());
        }
    }

    @ReactMethod
    public void setBeaconHeaders(ReadableMap headers, String platform){
        if(this.shouldWorkOnAndroid(platform)){
            if (headers == null) {
                Dynatrace.setBeaconHeaders(null);
            } else {
                Map<String, String> beaconHeaders = new HashMap<>();
                ReadableMapKeySetIterator iterator = headers.keySetIterator();
                while (iterator.hasNextKey()) {
                    String currentKey = iterator.nextKey();
                    beaconHeaders.put(currentKey, headers.getString(currentKey));
                }
                Dynatrace.setBeaconHeaders(beaconHeaders);
            } 
        }
    }
    
    @ReactMethod
    public void applyUserPrivacyOptions(ReadableMap userPrivacyOptions, String platform){
        if(this.shouldWorkOnAndroid(platform)){
            UserPrivacyOptions.Builder optionsBuilder = UserPrivacyOptions.builder();
            optionsBuilder.withCrashReportingOptedIn(userPrivacyOptions.getBoolean("_crashReportingOptedIn"));
            optionsBuilder.withDataCollectionLevel(DataCollectionLevel.valueOf(userPrivacyOptions.getString("_dataCollectionLevel")));

            Dynatrace.applyUserPrivacyOptions(optionsBuilder
                .build()
            );
        }
    }

    @ReactMethod
    public void getUserPrivacyOptions(String platform, Promise promise){
        if(this.shouldWorkOnAndroid(platform)){
            UserPrivacyOptions options = Dynatrace.getUserPrivacyOptions();
            WritableMap privacyMap = Arguments.createMap();
            privacyMap.putString("dataCollectionLevel", String.valueOf(options.getDataCollectionLevel()));
            privacyMap.putBoolean("crashReportingOptedIn", options.isCrashReportingOptedIn());
            promise.resolve(privacyMap);
        }
    }

    private void newAction(String name, String key) throws Exception {
        if (name == null) throw new Exception("action name is null");
        actions.put(key, _internal.enterAction(name));
    }

    private DTXAction getAction(String key){
        return actions.get(key);
    }

    private Boolean shouldWorkOnAndroid(String platform) {
        return platform == null || platform.equals(PLATFORM_ANDROID) || platform.equals("");
    }

}

File: ios/DynatraceRNBridge.m
//
//  DynatraceRNBridge.m
//

#import "DynatraceRNBridge.h"
#import <Foundation/Foundation.h>
#import <CoreLocation/CoreLocation.h>

@implementation DynatraceRNBridge

NSMutableDictionary *actionDict;
NSMutableDictionary *webTimingsDict;

NSString *const PlatformAndroid = @"android";
NSString *const PlatformiOS = @"ios";
NSString *const DataCollectionOff = @"OFF";
NSString *const DataCollectionPerformance = @"PERFORMANCE";
NSString *const DataCollectionUserBehavior = @"USER_BEHAVIOR";

RCT_EXPORT_MODULE(DynatraceBridge);

- (id) init
{
  self = [super init];
  if (self) {
    actionDict = [[NSMutableDictionary alloc] init];
    webTimingsDict = [[NSMutableDictionary alloc] init];
  }
  return self;
}

- (NSDictionary *)constantsToExport
{
    return @{ @"PLATFORM_ANDROID" : PlatformAndroid,
              @"PLATFORM_IOS" : PlatformiOS,
              @"DATA_COLLECTION_OFF" : DataCollectionOff,
              @"DATA_COLLECTION_PERFORMANCE" : DataCollectionPerformance,
              @"DATA_COLLECTION_USERBEHAVIOR" : DataCollectionUserBehavior };
}

RCT_EXPORT_METHOD(start:(NSDictionary *) options)
{
    if (options == nil) {
        return;
    }
    
    NSMutableDictionary<NSString*, id> *properties = [[NSMutableDictionary alloc] init];
    
    if (options[@"applicationId"] != NULL) {
        properties[@"DTXApplicationID"] = options[@"applicationId"];
    }

    if (options[@"beaconUrl"] != NULL) {
        properties[@"DTXBeaconURL"] = options[@"beaconUrl"];
    }

    if (options[@"userOptIn"] != NULL && [[options valueForKey:@"userOptIn"] isEqual: @(YES)]) {
        properties[@"DTXUserOptIn"] = @YES;
    }
    
    if (options[@"reportCrash"] != NULL && [[options valueForKey:@"reportCrash"] isEqual: @(NO)]) {
        properties[@"DTXCrashReporting"] = @NO;
    }
    
    if (options[@"logLevel"] != NULL && [((NSNumber *) options[@"logLevel"]) intValue] == 0){
        properties[@"DTXLogLevel"] = @"ALL";
    }

    if (properties[@"DTXBeaconURL"] != NULL && properties[@"DTXApplicationID"] != NULL) {
        [Dynatrace startupWithConfig:properties];
    }
}

RCT_EXPORT_METHOD(enterAction:(NSString *)name key:(nonnull NSString *)key platform: (NSString *) platform)
{
  if ([self shouldWorkOnIosWithPlatform: platform])
  {
    [self newAction:name key:key parentAction:nil];
  }
}

RCT_EXPORT_METHOD(enterManualAction:(NSString *)name key:(nonnull NSString *)key platform: (NSString *) platform)
{
  if ([self shouldWorkOnIosWithPlatform: platform])
  {
    DTXAction *action = [DTXAction enterActionWithName:name];
    
    if (action)
    {
        [actionDict setObject:action forKey:key];
    }
  }
}

RCT_EXPORT_METHOD(enterManualActionWithParent:(NSString *)name key:(nonnull NSString *)key parentKey:(nonnull NSString *)parentKey platform: (NSString *) platform)
{
    if ([self shouldWorkOnIosWithPlatform: platform])
    {
        DTXAction *parentAction = [self getAction:parentKey];

        if (parentAction != nil)
        {
            DTXAction *childAction = [DTXAction enterActionWithName:name parentAction:parentAction];
            
            if (childAction)
            {
                [actionDict setObject:childAction forKey:key];
            }
        }
        else
        {
            [self enterManualAction:name key:key platform:platform];
        }
    }
}

RCT_EXPORT_METHOD(leaveAction:(nonnull NSString *)key leave: (BOOL) leave platform: (NSString *) platform)
{
    if ([self shouldWorkOnIosWithPlatform: platform])
    {
        DTXAction *action = [self getAction:key];
        if (action == nil) return;
        [actionDict removeObjectForKey:key];

        if(leave)
        {
            [action leaveAction];
        }
    }
}

RCT_EXPORT_METHOD(cancelAction:(nonnull NSString *)key platform: (NSString *) platform)
{
    if ([self shouldWorkOnIosWithPlatform: platform])
    {
        DTXAction *action = [self getAction:key];
        if (action == nil) return;
        [actionDict removeObjectForKey:key];
        [action cancelAction];
    }
}

RCT_EXPORT_METHOD(endVisit: (NSString *) platform)
{
    if ([self shouldWorkOnIosWithPlatform: platform])
    {
        [Dynatrace endVisit];
    }
}

RCT_EXPORT_METHOD(reportErrorWithoutStacktrace:(NSString *)errorName errorCode:(nonnull NSNumber *)errorCode platform: (NSString *) platform)
{
    if ([self shouldWorkOnIosWithPlatform: platform])
    {
        DTXAction *action = [DTXAction enterActionWithName:@"Error"];
        [action reportErrorWithName:errorName errorValue:[errorCode intValue]];
        [action leaveAction];
    }
}

RCT_EXPORT_METHOD(reportError:(NSString *)errorName errorValue:(NSString *)errorValue errorReason:(NSString *)errorReason stacktrace:(NSString *)stacktrace platform: (NSString *) platform)
{
    if ([self shouldWorkOnIosWithPlatform: platform])
    {
        [DTXAction reportExternalErrorForPlatformType:-1 errorName:errorName errorValue:errorValue reason:errorReason stacktrace:stacktrace];
    }
}

RCT_EXPORT_METHOD(reportCrash:(NSString *)errorName errorReason:(NSString *)errorReason stacktrace:(NSString *)stacktrace isRealError:(BOOL)isRealError newSession:(BOOL)newSession platform: (NSString *) platform)
{
    if ([self shouldWorkOnIosWithPlatform: platform])
    {
        if(isRealError){
            [DTXAction reportExternalCrashForPlatformType:DTXActionPlatformJavaScript crashName:errorName reason:errorReason stacktrace:stacktrace];
        }else{
            [DTXAction reportExternalCrashForPlatformType:-1 crashName:errorName reason:errorReason stacktrace:stacktrace];
        }
        
        // We ignore the newSession parameter and always end the session as the iOS agent has no troubles with this behavior
        [Dynatrace endVisit];
    }
}

RCT_EXPORT_METHOD(reportErrorInAction:(nonnull NSString *)key errorName:(NSString *)errorName errorCode:(nonnull NSNumber *)errorCode platform: (NSString *) platform)
{
    if ([self shouldWorkOnIosWithPlatform: platform])
    {
        DTXAction *action = [self getAction:key];
        if (action == nil)
            return;
        [action reportErrorWithName:errorName errorValue:[errorCode intValue]];
    }
}

RCT_EXPORT_METHOD(reportStringValueInAction:(nonnull NSString *)actionKey withName:(NSString *)name value: (NSString *)value platform: (NSString *) platform)
{
  if ([self shouldWorkOnIosWithPlatform: platform])
  {
    DTXAction *action = [self getAction:actionKey];
    if (action == nil) return;
    [action reportValueWithName:name stringValue:value];
  }
}

RCT_EXPORT_METHOD(reportIntValueInAction:(nonnull NSString *)actionKey withName:(NSString *)name value: (nonnull NSNumber *)value platform: (NSString *) platform)
{
  if ([self shouldWorkOnIosWithPlatform: platform])
  {
    DTXAction *action = [self getAction:actionKey];
    if (action == nil) return;
    [action reportValueWithName:name intValue:value.intValue];
  }
}

RCT_EXPORT_METHOD(reportDoubleValueInAction:(nonnull NSString *)actionKey withName:(NSString *)name value: (nonnull NSNumber *)value platform: (NSString *) platform)
{
  if ([self shouldWorkOnIosWithPlatform: platform])
  {
    DTXAction *action = [self getAction:actionKey];
    if (action == nil) return;
    [action reportValueWithName:name doubleValue:value.doubleValue];
  }
}

RCT_EXPORT_METHOD(getRequestTag:(nonnull NSString *)actionKey withUrl:(NSString *)url findEventsWithResolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
    DTXAction* action = [self getAction:actionKey];
    resolve([action getTagForURL:[NSURL URLWithString:url]]);
}

RCT_EXPORT_METHOD(startWebRequestTiming:(NSString*) requestTag url:(NSString*) url)
{
    if(requestTag != NULL && url != NULL){
        DTXWebRequestTiming* timing = [DTXWebRequestTiming getDTXWebRequestTiming:requestTag requestUrl:[NSURL URLWithString:url]];
        if (timing != NULL) {
            [webTimingsDict setObject:timing forKey:[NSString stringWithString:requestTag]];
            [timing startWebRequestTiming];
        }
    }
}

RCT_EXPORT_METHOD(stopWebRequestTiming:(NSString*) requestTag url:(NSString*)url responseCode:(nonnull NSNumber*) responseCode responseMessage:(NSString*)responseMessage)
{
    if(requestTag != NULL){
        DTXWebRequestTiming* timing = [webTimingsDict objectForKey:requestTag];
        if(timing){
            [timing stopWebRequestTiming:[responseCode stringValue]];
            [webTimingsDict removeObjectForKey:requestTag];
        }
    }
}

RCT_EXPORT_METHOD(identifyUser:(NSString *)user platform: (NSString *) platform)
{
    if ([self shouldWorkOnIosWithPlatform: platform])
    {
        [Dynatrace identifyUser:user];
    }
}

RCT_EXPORT_METHOD(reportEventInAction:(nonnull NSString *)actionKey withName: (NSString *)name platform: (NSString *) platform)
{
    if ([self shouldWorkOnIosWithPlatform: platform])
    {
        DTXAction *action = [self getAction:actionKey];
        if (action == nil) return;
        [action reportEventWithName: name];
    }
}

RCT_EXPORT_METHOD(sendEvent:(NSString *)name withAttributes:(NSDictionary<NSString*, id>*) attributes platform: (NSString *) platform)
{
    if ([self shouldWorkOnIosWithPlatform: platform])
    {
        [Dynatrace sendEventWithName:name attributes:attributes];
    }
}

RCT_EXPORT_METHOD(sendBizEvent:(NSString *)type withAttributes:(NSDictionary<NSString*, id>*) attributes platform: (NSString *) platform)
{
    if ([self shouldWorkOnIosWithPlatform: platform])
    {
        [Dynatrace sendBizEventWithType:type attributes:attributes];
    }
}

RCT_EXPORT_METHOD(setGPSLocation:(double)latitude andLongitude: (double)longitude platform: (NSString *) platform)
{
    if ([self shouldWorkOnIosWithPlatform: platform])
    {
        CLLocation *location = [[CLLocation alloc] initWithLatitude:latitude longitude:longitude];
        [Dynatrace setGpsLocation:location];
    }
}

RCT_EXPORT_METHOD(flushEvents:(NSString *) platform)
{
    if ([self shouldWorkOnIosWithPlatform: platform])
    {
        [Dynatrace flushEvents];
    }
}

RCT_EXPORT_METHOD(isCrashReportingOptedIn:(NSString *) platform findEventsWithResolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
    if ([self shouldWorkOnIosWithPlatform: platform])
    {
        if([Dynatrace crashReportingOptedIn]){
            resolve(@"true");
        }else{
            resolve(@"false");
        }
    }
}

RCT_EXPORT_METHOD(setCrashReportingOptedIn:(BOOL) crashReportingOptedIn platform: (NSString *) platform)
{
    if ([self shouldWorkOnIosWithPlatform: platform])
    {
        [Dynatrace setCrashReportingOptedIn:crashReportingOptedIn];
    }
}

RCT_EXPORT_METHOD(getDataCollectionLevel:(NSString *) platform findEventsWithResolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
    if ([self shouldWorkOnIosWithPlatform: platform])
    {
        int i = [Dynatrace dataCollectionLevel];
        
        if(i == DTX_DataCollectionUserBehavior){
            resolve(DataCollectionUserBehavior);
        }else if(i == DTX_DataCollectionPerformance){
            resolve(DataCollectionPerformance);
        }else{
            resolve(DataCollectionOff);
        }
    }
}

RCT_EXPORT_METHOD(setDataCollectionLevel:(nonnull NSString *) dataCollectionLevel platform: (NSString *) platform)
{
    if ([self shouldWorkOnIosWithPlatform: platform])
    {
        int dataCollectionLevelEnum;
        
        if([dataCollectionLevel isEqualToString: DataCollectionPerformance]){
            dataCollectionLevelEnum = DTX_DataCollectionPerformance;
        }else if([dataCollectionLevel isEqualToString: DataCollectionUserBehavior]){
            dataCollectionLevelEnum = DTX_DataCollectionUserBehavior;
        }else{
            dataCollectionLevelEnum = DTX_DataCollectionOff;
        }
        
        [Dynatrace setDataCollectionLevel:dataCollectionLevelEnum completion:^(BOOL successful) {
            // Do nothing - Not sure
        }];
    }
}

RCT_EXPORT_METHOD(setBeaconHeaders:(NSDictionary *) headers platform: (NSString *) platform)
{
    if ([self shouldWorkOnIosWithPlatform: platform])
    {
        if (!headers) {
            [Dynatrace setBeaconHeaders:nil];
        } else {
            [Dynatrace setBeaconHeaders:headers];
        }
    }
}

RCT_EXPORT_METHOD(getUserPrivacyOptions:(NSString *) platform findEventsWithResolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
    if ([self shouldWorkOnIosWithPlatform: platform])
    {
        id<DTXUserPrivacyOptions> privacyConfig = [Dynatrace userPrivacyOptions];
        NSString * level;
        
        if(privacyConfig.dataCollectionLevel == DTX_DataCollectionUserBehavior){
            level = DataCollectionUserBehavior;
        }else if(privacyConfig.dataCollectionLevel == DTX_DataCollectionPerformance){
            level = DataCollectionPerformance;
        }else{
            level = DataCollectionOff;
        }

        NSDictionary *privacyDict = @{@"dataCollectionLevel": level, @"crashReportingOptedIn": [NSNumber numberWithBool: privacyConfig.crashReportingOptedIn]};

        resolve(privacyDict);
    }
}

RCT_EXPORT_METHOD(applyUserPrivacyOptions:(NSDictionary *) userPrivacyOptions platform: (NSString *) platform)
{
    if ([self shouldWorkOnIosWithPlatform: platform])
    {
        id<DTXUserPrivacyOptions> privacyConfig = [Dynatrace userPrivacyOptions];
        
        if([[userPrivacyOptions valueForKey:@"_dataCollectionLevel"] isEqualToString: DataCollectionPerformance]){
            privacyConfig.dataCollectionLevel = DTX_DataCollectionPerformance;
        }else if([[userPrivacyOptions valueForKey:@"_dataCollectionLevel"] isEqualToString: DataCollectionUserBehavior]){
            privacyConfig.dataCollectionLevel = DTX_DataCollectionUserBehavior;
        }else if([[userPrivacyOptions valueForKey:@"_dataCollectionLevel"] isEqualToString: DataCollectionOff]){
            privacyConfig.dataCollectionLevel = DTX_DataCollectionOff;
        } else {
            // do nothing and keep current value
        }
        
        if ([[userPrivacyOptions valueForKey:@"_crashReportingOptedIn"] isEqual: @(YES)]) {
            privacyConfig.crashReportingOptedIn = YES;
        } else if ([[userPrivacyOptions valueForKey:@"_crashReportingOptedIn"]  isEqual: @(NO)]) {
            privacyConfig.crashReportingOptedIn = NO;
        } else {
            // do nothing and keep current value
        }

        [Dynatrace applyUserPrivacyOptions:privacyConfig completion:^(BOOL successful) {
            // do nothing with callback
        }];
    }
}

- (void) newAction:(NSString *)name key:(nonnull NSString *)key parentAction:(DTXAction *)parentAction
{
    DTXAction *action = [DTXAction integrateActionWithName:name];
    
    if (action)
    {
        [actionDict setObject:action forKey:key];
    }
}

- (DTXAction *) getAction:(nonnull NSString *)key
{
  return [actionDict objectForKey:key];
}


+ (BOOL)requiresMainQueueSetup
{
  return YES;
}

- (BOOL) shouldWorkOnIosWithPlatform: (NSString *) platform
{
    return platform == nil || [platform isEqualToString:PlatformiOS] || [platform isEqualToString:@""];
}

@end